kpops/components/base_components/models/from_section.py

```diff
@@ -16,26 +16,53 @@ class InputTopicTypes(str, Enum):
 
     INPUT = "input"
     EXTRA = "extra"
+    PATTERN = "pattern"
     INPUT_PATTERN = "input-pattern"
     EXTRA_PATTERN = "extra-pattern"
 
 
 class FromTopic(BaseModel):
     """Input topic
 
-    :param type: Topic type
-    :type type: InputTopicTypes
-    :param role: Custom identifier belonging to a topic, provide only if `type` is `extra` or `extra-pattern`
-    :type role: str | None
+    :param type: Topic type, defaults to None
+    :type type: InputTopicTypes | None, optional
+    :param role: Custom identifier belonging to a topic, provide only if `type` is `extra` or `extra-pattern`.
+        When `role` is not `None`, `type: pattern` is equal to `type: extra-pattern`, defaults to None
+    :type role: str | None, optional
     """
 
-    type: InputTopicTypes = Field(..., description=describe_attr("type", __doc__))
+    type: InputTopicTypes | None = Field(
+        default=None, description=describe_attr("type", __doc__)
+    )
     role: str | None = Field(default=None, description=describe_attr("role", __doc__))
 
     class Config(DescConfig):
         extra = Extra.forbid
         use_enum_values = True
 
+    def __init__(self, **kwargs):
+        kwargs["type"] = self.__assign_type(
+            kwargs.get("type", None), kwargs.get("role", None)
+        )
+        super().__init__(**kwargs)
+
+    @staticmethod
+    def __assign_type(
+        type_: InputTopicTypes | None, role: str | None
+    ) -> InputTopicTypes:
+        match type_, role:
+            case None, None:
+                type_ = InputTopicTypes.INPUT
+            case None, _:
+                type_ = InputTopicTypes.EXTRA
+            case InputTopicTypes.PATTERN, None:
+                type_ = InputTopicTypes.INPUT_PATTERN
+            case InputTopicTypes.PATTERN, _:
+                type_ = InputTopicTypes.EXTRA_PATTERN
+            case _, _:
+                return type_  # type: ignore[return-value]
```

Comment:
Had to silence mypy here, it was concerned about returning `Optional[InputTopicTypes]` instead of `InputTopicTypes`, which I'm pretty sure is not the case as the cases narrow the type down.

Reply:
I think it's because you're reassigning `type_` instead of directly returning

Reply:
Didn't fix it unfortunately

Reply:
ok :( then it's not doing such a good job of type narrowing. Pyright gets it

Reply:
perhaps it's the same problem that we encountered here
https://github.com/bakdata/kpops/pull/197#discussion_r1177656235

Reply:
perhaps we should replace MyPy with Pyright in the CI, it seems a lot more powerful

Reply:
I read a bit and the MyPy vs Pyright comparison seems a bit one-sided in favor of the latter. I support the replacement. Should I create an issue?

Reply:
#250